packages/eslint-plugin/src/rules/no-deprecated.ts

```diff
@@ -0,0 +1,243 @@
+
+    function isDeclaration(node: IdentifierLike): boolean {
+      const { parent } = node;
+
+      switch (parent.type) {
+        case AST_NODE_TYPES.ArrayPattern:
+          return parent.elements.includes(node as TSESTree.Identifier);
+
+        case AST_NODE_TYPES.ClassExpression:
+        case AST_NODE_TYPES.ClassDeclaration:
+        case AST_NODE_TYPES.VariableDeclarator:
+        case AST_NODE_TYPES.TSEnumMember:
+          return parent.id === node;
+
+        case AST_NODE_TYPES.MethodDefinition:
+        case AST_NODE_TYPES.PropertyDefinition:
+          return parent.key === node;
+
+        case AST_NODE_TYPES.Property:
+          return (
+            (parent.shorthand && parent.value === node) ||
+            parent.parent.type === AST_NODE_TYPES.ObjectExpression
+          );
+
+        case AST_NODE_TYPES.AssignmentPattern:
+          return (
+            parent.left === node &&
+            !(
+              parent.parent.type === AST_NODE_TYPES.Property &&
+              parent.parent.shorthand
+            )
+          );
+
+        case AST_NODE_TYPES.ArrowFunctionExpression:
+        case AST_NODE_TYPES.FunctionDeclaration:
+        case AST_NODE_TYPES.FunctionExpression:
+        case AST_NODE_TYPES.TSDeclareFunction:
+        case AST_NODE_TYPES.TSEmptyBodyFunctionExpression:
+        case AST_NODE_TYPES.TSEnumDeclaration:
+        case AST_NODE_TYPES.TSInterfaceDeclaration:
+        case AST_NODE_TYPES.TSMethodSignature:
+        case AST_NODE_TYPES.TSModuleDeclaration:
+        case AST_NODE_TYPES.TSParameterProperty:
+        case AST_NODE_TYPES.TSPropertySignature:
+        case AST_NODE_TYPES.TSTypeAliasDeclaration:
+        case AST_NODE_TYPES.TSTypeParameter:
+          return true;
+
+        default:
+          return false;
+      }
+    }
+
+    function isInsideExportOrImport(node: TSESTree.Node): boolean {
+      let current = node;
+
+      while (true) {
+        switch (current.type) {
+          case AST_NODE_TYPES.ExportAllDeclaration:
+          case AST_NODE_TYPES.ExportDefaultDeclaration:
+          case AST_NODE_TYPES.ExportNamedDeclaration:
+          case AST_NODE_TYPES.ImportDeclaration:
+          case AST_NODE_TYPES.ImportExpression:
+            return true;
+
+          case AST_NODE_TYPES.ArrowFunctionExpression:
+          case AST_NODE_TYPES.BlockStatement:
+          case AST_NODE_TYPES.ClassBody:
+          case AST_NODE_TYPES.TSInterfaceDeclaration:
+          case AST_NODE_TYPES.FunctionDeclaration:
+          case AST_NODE_TYPES.FunctionExpression:
+          case AST_NODE_TYPES.Program:
+          case AST_NODE_TYPES.TSUnionType:
+          case AST_NODE_TYPES.VariableDeclarator:
+            return false;
+
+          default:
+            current = current.parent;
+        }
+      }
+    }
+
+    function getJsDocDeprecation(
+      symbol: ts.Signature | ts.Symbol | undefined,
+    ): string | undefined {
+      const tag = symbol
+        ?.getJsDocTags(checker)
```

Comment:
this forces us to do "jsdoc parsing" by default :(
I kind of wish we could do it without using this -- eg with a regex so we can speed up parsing.

Reply:
_OH_, this just reminded me we have `JSDocParsingMode`. Adding `jsDocParsingMode: 'none'` to the test file's `languageOptions.parserOptions` causes all the `invalid` tests to fail for not having any parsed JSDocs. Great.

I think we're going to have to do our own regex-based parsing. I'll put the PR back in draft in the meantime.

Reply:
I think this is one of the few usecases of jsdoc via TS AST in the ecosystem. So yeah if we can remove this then we can consider flipping the default in the next version to speed up parsing in the general case!

Reply:
Well, I tried using on `symbol.valueDeclaration` and `symbol.valueDeclaration.parent` in the first `invalid` test case:
* `ts.getJSDocCommentsAndTags`: Nothing retrieved
* `tsutils.forEachComment`: Retrieved the comment, but took ~750ms 

`deprecation/deprecation` also uses `symbol.getJsDocTags` and suffers from the same issue. We don't (yet?) default `jsdocParsingMode` to `'none'` or `'type-info'` for lint users. I added a thrown error for those modes in 5cd8b823d.

Back-linking to past conversation (cc @jakebailey as FYI): https://github.com/typescript-eslint/typescript-eslint/pull/7997#issuecomment-1828891453

Reply:
Yeah for sure - file an issue.
Wasnt sposed to be a blocking comment. More of a "hey if we can that'd be good" thing

Reply:
Perhaps it's worth adding `@deprecated` to the list of tags that we parse out unconditionally in the API? Would that be simpler?

Reply:
I'd have to test and see. Unfortunate though as my enum names were too clever, `@deprecated` is not used for checking :(